### ISSUE TITLE
Connecting to authenticated AWS AppSync GraphQL endpoint

### ISSUE THREAD

**Issue by jakeherp** [gatsbyjs/gatsby#18684]
I am trying to connect a Gatsby site using `gatsby-source-graphql` to an AppSync endpoint, that requires authentication and needs to be signed.
My approach so far was to pull in `sign` from the aws4 package and sign it as follows:

```js
require("dotenv").config({
  path: `.env`,
});
const { sign } = require('aws4');

const { ACCESS_KEY_ID, ACCESS_SECRET_KEY, API_ENDPOINT } = { ...process.env };
const url = new URL(API_ENDPOINT);

module.exports = {
  plugins: [
    ...
    {
      resolve: `gatsby-source-graphql`,
      options: {
        typeName: "AWSAppSync",
        fieldName: "mygql",
        url: API_ENDPOINT,
        fetch: () => {
          sign({
            service: 'appsync',
            region: 'eu-west-1',
            host: url.hostname,
            path: url.pathname,
            method: 'POST',
            body: JSON.stringify(params),
            headers: { 'Content-Type': 'application/json' },
          }, { accessKeyId: ACCESS_KEY_ID, secretAccessKey: ACCESS_SECRET_KEY })
        }
      },
    },
  ]
}
```

But this fails with a 403 response:

```sh
 ERROR #11321  PLUGIN

"gatsby-source-graphql" threw an error while running the sourceNodes lifecycle:

Response not successful: Received status code 403

  ServerError: Response not successful: Received status code 403
  
  - index.js:23 Object.exports.throwServerError
    [gatsby-site]/[apollo-link-http-common]/lib/index.js:23:17
  
  - index.js:48 
    [gatsby-site]/[apollo-link-http-common]/lib/index.js:48:21
  
  - next_tick.js:68 process._tickCallback
    internal/process/next_tick.js:68:7
```

I assume that the `fetch` method for the plugin doesn't do what I expected, but how else would I pass this information in?

**Comment by pieh** [comment]
I don't have much insight on this particular issue, but saw that we had few similar issues already ( https://github.com/gatsbyjs/gatsby/issues?q=is%3Aissue+sort%3Aupdated-desc+AWS+AppSync+GraphQL+is%3Aclosed ). Comments in this https://github.com/gatsbyjs/gatsby/issues/17302 issue seems to suggest that this is solvable?

> I assume that the `fetch` method for the plugin doesn't do what I expected, but how else would I pass this information in?

`gatsby-source-graphql` doesn't support `fetch` config option, we do have `fetchOptions` and `createLink`. It seems like the way you use it, is not doing anything?

**Comment by vladar** [comment]
Published in `gatsby-source-graphql 2.1.24`

**Comment by multidis** [comment]
@vladar could you please clarify what exactly was implemented in `gatsby-source-graphql 2.1.24`?

How can we move from [this API key example](https://github.com/aws-samples/aws-appsync-gatsby-sample/blob/master/gatsby-config.js#L13) to using `AWS_IAM` or Cognito auth [with Amplify](https://github.com/gatsbyjs/gatsby/issues/17302#issuecomment-544110352)? Thanks!

**Comment by vladar** [comment]
This PR was published: #19630

**Comment by multidis** [comment]
I see now, a custom `node-fetch` call works. In case someone is looking for an example `gatsby-config.js` modified based on @jakeherp 's code above:

```javascript
const { sign } = require('aws4');
const fetch = require(`node-fetch`);

const { API_ENDPOINT_ID, API_REGION, ACCESS_KEY_ID, ACCESS_SECRET_KEY } = { ...process.env };
const API_ENDPOINT = `https://${API_ENDPOINT_ID}.appsync-api.${API_REGION}.amazonaws.com/graphql`;
const url = new URL(API_ENDPOINT);

module.exports = {
    plugins: [
        {
            resolve: "gatsby-source-graphql",
            options: {
                typeName: "AWSAppSync",
                fieldName: "events",
                url: API_ENDPOINT,
                fetch: (uri, options = {}) => fetch(uri, sign({ ...options,
                      service: 'appsync',
                      region: API_REGION,
                      host: url.hostname,
                      path: url.pathname,
                      method: 'POST',
                      headers: { 'Content-Type': 'application/json' },
                  },
                  { accessKeyId: ACCESS_KEY_ID, secretAccessKey: ACCESS_SECRET_KEY }
                  ))
            },
        },
    ],
}
```

The user whose credentials are supplied via env should have appropriate AppSync permissions, for example this policy works for reading:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "appsync:GraphQL",
                "appsync:List*",
                "appsync:Get*"
            ],
            "Resource": [
                "arn:aws:appsync:{region}:{accountId}:apis/{apiId}/*"
            ]
        }
    ]
}
```
